File: risc0/zkp/core/test/fp.cpp
// Copyright 2022 Risc0, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "risc0/zkp/core/fp.h"
#include "risc0/core/rng.h"

#include <gtest/gtest.h>

namespace risc0 {

// Compare core operations against simple % P implementations
TEST(Fp, FpCompareNative) {
  PsuedoRng rng(2);
  for (size_t i = 0; i < 100000; i++) {
    Fp fa = Fp::random(rng);
    Fp fb = Fp::random(rng);
    uint64_t a = fa.asUInt32();
    uint64_t b = fb.asUInt32();
    EXPECT_EQ(fa + fb, Fp((a + b) % Fp::P));
    EXPECT_EQ(fa - fb, Fp((a + (Fp::P - b)) % Fp::P));
    EXPECT_EQ(fa * fb, Fp((a * b) % Fp::P));
  }
}

// Smoke tests for pow
TEST(Fp, FpTestPow) {
  EXPECT_EQ(pow(Fp(5), 0), 1);
  EXPECT_EQ(pow(Fp(5), 1), 5);
  EXPECT_EQ(pow(Fp(5), 2), 25);
  // Mathematica says PowerMod[5, 1000, 15*2^27 + 1] == 589699054
  EXPECT_EQ(pow(Fp(5), 1000), 589699054);
  EXPECT_EQ(pow(Fp(5), Fp::P - 2) * Fp(5), 1);
  EXPECT_EQ(pow(Fp(5), Fp::P - 1), 1);
}

// Smoke test for inv
TEST(Fp, FpTestInv) {
  EXPECT_EQ(inv(Fp(5)) * Fp(5), 1);
}

} // End namespace risc0

File: risc0/circuit/rv32im-sys/src/fp.rs
// Copyright 2026 RISC Zero, Inc.
//
// Licensed under the Apache License, Version 2.0, <LICENSE-APACHE or
// http://apache.org/licenses/LICENSE-2.0> or the MIT license <LICENSE-MIT or
// http://opensource.org/licenses/MIT>, at your option. This file may not be
// copied, modified, or distributed except according to those terms.
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// SPDX-License-Identifier: Apache-2.0 OR MIT

use bytemuck::{Pod, Zeroable};
use risc0_zkp::field::baby_bear::{self, Elem as BabyBearElem};

/// This is the field point type that proving uses. It has weaker guarantees than BabyBearElem and
/// implements `Pod`
#[derive(Debug, Copy, Clone, PartialEq, Eq, Pod, Zeroable, Default, PartialOrd, Ord)]
#[repr(C)]
pub struct Fp(u32);

impl From<BabyBearElem> for Fp {
    fn from(elem: BabyBearElem) -> Self {
        Self(elem.as_u32())
    }
}

impl Fp {
    pub fn as_raw(&self) -> u32 {
        self.0
    }

    pub fn new_raw(v: u32) -> Self {
        Self(v)
    }

    pub fn from_elem_array<const N: usize>(elems: [BabyBearElem; N]) -> [Self; N] {
        let mut self_ = [Self(0); N];
        for (self_cell, &elem) in self_.iter_mut().zip(elems.iter()) {
            *self_cell = Self(elem.as_u32_montgomery());
        }
        self_
    }

    pub fn as_elem_array<const N: usize>(elems: &[Self; N]) -> &[BabyBearElem; N] {
        for &elem in elems {
            assert!(elem.0 < baby_bear::P, "invalid baby bear element");
        }
        unsafe { std::mem::transmute(elems) }
    }

    pub fn to_elem(&self) -> BabyBearElem {
        assert!(self.0 < baby_bear::P, "invalid baby bear element");
        BabyBearElem::new_raw(self.0)
    }
}
